src/components/enterprise-banner/OffersAlert.jsx

```diff
@@ -0,0 +1,57 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import { Alert } from '@edx/paragon';
+
+import { fetchOffers } from '../dashboard/sidebar/offers';
+
+export const getOffersText = (number) => `You have ${number} offers available.`;
+
+export class BaseOffersAlert extends React.Component {
+  componentDidMount() {
+    const { isOffersLoading, fetchOffersAction } = this.props;
+
+    if (!isOffersLoading) {
+      fetchOffersAction('full_discount_only=True');
+    }
+  }
+
+  render() {
+    const { offersCount } = this.props;
+    return (
+      <>{offersCount && (
+        <Alert
+          className="pl-5"
+          variant="info"
+        >
+          <div className="container">{getOffersText(offersCount)}</div>
+        </Alert>
+      )}
+      </>
+    );
+  }
+}
+
+BaseOffersAlert.defaultProps = {
+  fetchOffersAction: null,
+  isOffersLoading: false,
+  offersCount: 0,
+};
+
+BaseOffersAlert.propTypes = {
+  fetchOffersAction: PropTypes.func,
+  isOffersLoading: PropTypes.bool,
+  offersCount: PropTypes.number,
+};
+
+const mapStateToProps = state => ({
+  isOffersLoading: state.offers.loading,
+  offersCount: state.offers.offersCount,
+  timeFetched: state.offers.timeFetched,
+});
+
+const mapDispatchToProps = dispatch => ({
+  fetchOffersAction: (query) => (query ? dispatch(fetchOffers(query)) : dispatch(fetchOffers())),
+});
+
+export default connect(mapStateToProps, mapDispatchToProps)(BaseOffersAlert);
```

Comment:
I'm trying to grasp what this is doing (I have seen these dispatches and connectors before, but not super clear on them). 
So is this fetching data and recording it as props which then map to state?
Or is it taking props and using them to fetch data and then mapping the result to state?

I'm just trying to grasp the use of props here since I tend to associate those with coming from a parent component not after a call in the component.

Reply:
Cool! So, `connect` is what React calls a Higher Order Component (HoC). https://reactjs.org/docs/higher-order-components.html (I've never fount that doc super useful, but maybe you will?)
It basically wraps the component and returns another one, which allows a bunch of otherwise unrelated components to have access to the same Redux store, where we keep a bunch of data. https://react-redux.js.org/api/connect

I think the easiest way to think of it is that `mapStateToProps` is  a  **getter** function. It gives you access to the Redux store, and you pass various data from the store to your component as props. That way, in your component, things can have straightforward names and you're not digging through the whole store. `mapStateToProps` occasionally has some data processing in it.

`mapDispatchToProps`, on the other hand, gives you access to **setter** functions, namely `dispatch`, which is a function that takes an action and passes it to the reducer, which is the only way to modify what Redux has in the store (which is an immutable object). So here we're defining a setter that dispatch various actions to the store (which were already written, so I didn't have to define them in this PR.) It will send an action to the store to let Redux know that the request was made, giving us the `isLoading` state, and then when data (or an error) is received, the store is updated. The action, in this case, `fetchOffers`, is also passed to my component as a prop. (Which incidentally, is why I'm exporting it as an **unconnected** component for testing; I can then pass in a spy as the fetchOffersAction to test that it is called, etc.)

Let me know if this makes no sense or if you want more detail!

Reply:
Thanks! Mostly makes sense. It sounds similar to using contexts but in a bit of a different way.

Reply:
Yeah, totally. In fact, I basically think we should drop Redux entirely in favor of React context. (Historically, react context was kind of terrible, but now it's not, so...)
But this was already set up to use the Redux store, so I took advantage of existing code.

Reply:
Gotcha! Okay that all makes sense now. I had learned React only about a year ago when context already existed so I entirely missed Redux which is why it probably confuses me 😅